File: WaterCue/Services/GroqVisionService.swift
import Foundation
import AppKit

struct ValidationResult {
    let valid: Bool
    let reason: String
}

enum GroqError: LocalizedError {
    case noAPIKey
    case unauthorized
    case rateLimited
    case serverError(Int)
    case networkError(Error)
    case timeout
    case unparseable(String)
    case replayedPhoto

    var errorDescription: String? {
        switch self {
        case .noAPIKey:         return "Configure sua Groq API key em Preferências."
        case .unauthorized:     return "API key inválida. Verifique em Preferências."
        case .rateLimited:      return "Servidor ocupado. Tente novamente em instantes."
        case .serverError(let c): return "Erro no servidor Groq (\(c)). Tente novamente."
        case .networkError:     return "Sem conexão. Verifique o Wi-Fi."
        case .timeout:          return "Groq demorou muito. Tente novamente."
        case .unparseable:      return "Resposta inesperada. Tente outra foto."
        case .replayedPhoto:    return "Essa foto já foi usada recentemente. Tire uma foto nova."
        }
    }
}

actor GroqVisionService {
    static let shared = GroqVisionService()

    private let baseURL = URL(string: "https://api.groq.com/openai/v1/chat/completions")!
    private let session: URLSession = {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 15
        config.timeoutIntervalForResource = 20
        return URLSession(configuration: config)
    }()

    private let systemPrompt = "Você é um validador de hidratação. Responda APENAS JSON válido no formato {\"valid\":bool,\"reason\":string}. Não escreva nada antes ou depois do JSON."

    private let userPrompt = "Analise a imagem. valid=true se: há uma pessoa ou mão segurando ou tocando um recipiente de bebida (copo, caneca, garrafa, squeeze) — pode estar vazio, pode ter qualquer líquido, o que importa é o recipiente estar presente e a pessoa interagindo com ele. valid=false apenas se: não há recipiente visível, a imagem é claramente uma foto de tela ou foto antiga não tirada ao vivo, ou não há nenhuma pessoa/mão na imagem. reason: justifique em 1 frase em português."

    func validate(jpegData: Data, model: String) async throws -> ValidationResult {
        let key = try KeychainService.shared.getGroqKey()
        let resized = try resize(jpegData, maxDim: 1024)
        let b64 = resized.base64EncodedString()

        let body = GroqRequestBody(
            model: model,
            messages: [
                .init(role: "system", content: .text(systemPrompt)),
                .init(role: "user", content: .parts([
                    .init(type: "image_url", imageURL: .init(url: "data:image/jpeg;base64,\(b64)")),
                    .init(type: "text", text: userPrompt)
                ]))
            ],
            temperature: 0.1,
            maxTokens: 150,
            responseFormat: .init(type: "json_object")
        )

        var req = URLRequest(url: baseURL)
        req.httpMethod = "POST"
        req.setValue("Bearer \(key)", forHTTPHeaderField: "Authorization")
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONEncoder().encode(body)

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: req)
        } catch let urlError as URLError where urlError.code == .timedOut {
            throw GroqError.timeout
        } catch {
            throw GroqError.networkError(error)
        }

        guard let http = response as? HTTPURLResponse else {
            throw GroqError.unparseable("no http response")
        }
        switch http.statusCode {
        case 200: break
        case 401: throw GroqError.unauthorized
        case 429: throw GroqError.rateLimited
        default:  throw GroqError.serverError(http.statusCode)
        }

        return try parseResponse(data)
    }

    private func parseResponse(_ data: Data) throws -> ValidationResult {
        struct Choice: Decodable {
            struct Message: Decodable { let content: String }
            let message: Message
        }
        struct GroqResponse: Decodable { let choices: [Choice] }
        struct Inner: Decodable { let valid: Bool; let reason: String }

        let outer = try JSONDecoder().decode(GroqResponse.self, from: data)
        guard let content = outer.choices.first?.message.content else {
            throw GroqError.unparseable("no choices")
        }

        let innerData = content.data(using: .utf8) ?? Data()
        if let inner = try? JSONDecoder().decode(Inner.self, from: innerData) {
            return ValidationResult(valid: inner.valid, reason: inner.reason)
        }

        // fallback: find JSON object in text
        if let range = content.range(of: #"\{[^}]+\}"#, options: .regularExpression),
           let extracted = content[range].data(using: .utf8),
           let inner = try? JSONDecoder().decode(Inner.self, from: extracted) {
            return ValidationResult(valid: inner.valid, reason: inner.reason)
        }

        throw GroqError.unparseable(content)
    }

    private func resize(_ data: Data, maxDim: CGFloat) throws -> Data {
        guard let image = NSImage(data: data) else {
            throw GroqError.unparseable("cannot decode image")
        }
        let original = image.size
        let scale = min(maxDim / original.width, maxDim / original.height, 1.0)
        if scale >= 1.0 {
            return data
        }
        let newSize = NSSize(width: original.width * scale, height: original.height * scale)
        let resized = NSImage(size: newSize)
        resized.lockFocus()
        image.draw(in: NSRect(origin: .zero, size: newSize),
                   from: NSRect(origin: .zero, size: original),
                   operation: .copy, fraction: 1.0)
        resized.unlockFocus()

        guard let tiff = resized.tiffRepresentation,
              let bitmap = NSBitmapImageRep(data: tiff),
              let jpeg = bitmap.representation(using: .jpeg, properties: [.compressionFactor: 0.7]) else {
            return data
        }
        return jpeg
    }
}

// MARK: - Request types

private struct GroqRequestBody: Encodable {
    let model: String
    let messages: [Message]
    let temperature: Double
    let maxTokens: Int
    let responseFormat: ResponseFormat

    enum CodingKeys: String, CodingKey {
        case model, messages, temperature
        case maxTokens = "max_tokens"
        case responseFormat = "response_format"
    }

    struct Message: Encodable {
        let role: String
        let content: ContentValue

        enum CodingKeys: String, CodingKey { case role, content }
    }

    enum ContentValue: Encodable {
        case text(String)
        case parts([ContentPart])

        func encode(to encoder: Encoder) throws {
            switch self {
            case .text(let s):
                var c = encoder.singleValueContainer()
                try c.encode(s)
            case .parts(let p):
                var c = encoder.singleValueContainer()
                try c.encode(p)
            }
        }
    }

    struct ContentPart: Encodable {
        let type: String
        var imageURL: ImageURL?
        var text: String?

        enum CodingKeys: String, CodingKey {
            case type
            case imageURL = "image_url"
            case text
        }
    }

    struct ImageURL: Encodable {
        let url: String
    }

    struct ResponseFormat: Encodable {
        let type: String
    }
}

File: WaterCueWindows/Services/GroqVisionService.cs
using System.IO;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Windows.Media.Imaging;

namespace WaterCueWindows.Services;

public record ValidationResult(bool Valid, string Reason);

public enum GroqErrorKind
{
    NoApiKey, Unauthorized, RateLimited, ServerError, NetworkError, Timeout, Unparseable, ReplayedPhoto
}

public class GroqException(GroqErrorKind error, string message) : Exception(message)
{
    public GroqErrorKind Error { get; } = error;
}

public class GroqVisionService
{
    public static readonly GroqVisionService Shared = new();

    private static readonly Uri BaseUrl = new("https://api.groq.com/openai/v1/chat/completions");

    private readonly HttpClient _http = new() { Timeout = TimeSpan.FromSeconds(15) };

    private const string SystemPrompt =
        "Você é um validador de hidratação. Responda APENAS JSON válido no formato " +
        "{\"valid\":bool,\"reason\":string}. Não escreva nada antes ou depois do JSON.";

    private const string UserPrompt = """
        Esta é uma captura de webcam ao vivo de um sistema de hidratação. Avalie DUAS condições obrigatórias:

        CONDIÇÃO 1 — ROSTO HUMANO: pelo menos um rosto humano real está parcialmente visível e identificável. Critério mínimo: olhos, nariz ou boca detectáveis com clareza. NÃO conta: mãos isoladas, corpo sem rosto, silhuetas, reflexos, bonecos, fotos de pessoas, avatares.

        CONDIÇÃO 2 — RECIPIENTE EM USO: um recipiente de bebida (copo, garrafa, caneca, squeeze, xícara, copo descartável, garrafa esportiva) está visível E a pessoa está interagindo — segurando, levando à boca, ou claramente prestes a beber. O recipiente pode estar vazio. O recipiente deve estar em primeiro plano ou visivelmente associado à pessoa, não apenas ao fundo.

        valid=true SOMENTE SE as duas condições forem satisfeitas simultaneamente.

        valid=false se qualquer um dos seguintes for verdadeiro:
        - Nenhum rosto humano identificável na imagem
        - Nenhum recipiente de bebida visível ou associado à pessoa
        - A imagem exibe uma tela (monitor, celular, TV mostrando outra imagem ou vídeo)
        - A imagem é foto impressa, foto de foto, ou imagem claramente não capturada ao vivo agora
        - A imagem está muito escura, borrada ou obstruída para avaliar qualquer condição
        - O recipiente contém comida sólida (prato, tigela de comida) e não bebida
        - Há dúvida razoável sobre qualquer condição — na dúvida, prefira false

        reason: escreva 1 frase curta em português com tom bem-humorado e leve. Use criatividade baseada no caso específico detectado. Exemplos de tom (não copie, crie variações):
        - Rosto presente mas sem recipiente → "Seu rosto é lindo, mas cadê a água?"
        - Recipiente presente mas sem rosto → "Ótimo copo! Mas preciso ver você também, não só ele."
        - Imagem escura/borrada → "Tá com medo da câmera? Aparece aí!"
        - Foto de tela detectada → "Tentou me enganar com printscreen? Não rola não."
        - Foto de foto → "Foto antiga não vale, precisa ser ao vivo!"
        - valid=true, pessoa bebendo → algo animado celebrando a hidratação
        - valid=true, pessoa segurando → algo incentivando a beber logo
        """;

    private GroqVisionService() { }

    public async Task<ValidationResult> ValidateAsync(byte[] jpegData, string model)
    {
        var apiKey = SettingsService.Shared.LoadApiKey();
        if (string.IsNullOrEmpty(apiKey))
            throw new GroqException(GroqErrorKind.NoApiKey, "Configure sua Groq API key em Preferências.");

        var resized = ResizeJpeg(jpegData, 1024);
        var b64 = Convert.ToBase64String(resized);

        var payload = new
        {
            model,
            messages = new object[]
            {
                new { role = "system", content = SystemPrompt },
                new
                {
                    role = "user",
                    content = new object[]
                    {
                        new { type = "image_url", image_url = new { url = $"data:image/jpeg;base64,{b64}" } },
                        new { type = "text", text = UserPrompt }
                    }
                }
            },
            temperature = 0.1,
            max_tokens = 150,
            response_format = new { type = "json_object" }
        };

        using var request = new HttpRequestMessage(HttpMethod.Post, BaseUrl);
        request.Headers.Add("Authorization", $"Bearer {apiKey}");
        request.Content = new StringContent(
            JsonSerializer.Serialize(payload), Encoding.UTF8, "application/json");

        HttpResponseMessage response;
        try
        {
            response = await _http.SendAsync(request);
        }
        catch (TaskCanceledException)
        {
            throw new GroqException(GroqErrorKind.Timeout, "Groq demorou muito. Tente novamente.");
        }
        catch (Exception ex)
        {
            throw new GroqException(GroqErrorKind.NetworkError, $"Sem conexão. Verifique o Wi-Fi. ({ex.Message})");
        }

        return response.StatusCode switch
        {
            System.Net.HttpStatusCode.OK => await ParseResponseAsync(response),
            System.Net.HttpStatusCode.Unauthorized =>
                throw new GroqException(GroqErrorKind.Unauthorized, "API key inválida. Verifique em Preferências."),
            System.Net.HttpStatusCode.TooManyRequests =>
                throw new GroqException(GroqErrorKind.RateLimited, "Servidor ocupado. Tente novamente em instantes."),
            var code =>
                throw new GroqException(GroqErrorKind.ServerError, $"Erro no servidor Groq ({(int)code}). Tente novamente.")
        };
    }

    private static async Task<ValidationResult> ParseResponseAsync(HttpResponseMessage response)
    {
        var json = await response.Content.ReadAsStringAsync();
        try
        {
            using var doc = JsonDocument.Parse(json);
            var content = doc.RootElement
                .GetProperty("choices")[0]
                .GetProperty("message")
                .GetProperty("content")
                .GetString() ?? "";

            using var inner = JsonDocument.Parse(content);
            var valid = inner.RootElement.GetProperty("valid").GetBoolean();
            var reason = inner.RootElement.GetProperty("reason").GetString() ?? "";
            return new ValidationResult(valid, reason);
        }
        catch
        {
            throw new GroqException(GroqErrorKind.Unparseable, "Resposta inesperada. Tente outra foto.");
        }
    }

    private static byte[] ResizeJpeg(byte[] jpegData, int maxDim)
    {
        using var ms = new MemoryStream(jpegData);
        var frame = BitmapFrame.Create(ms, BitmapCreateOptions.None, BitmapCacheOption.OnLoad);
        var w = frame.PixelWidth;
        var h = frame.PixelHeight;
        var scale = Math.Min((double)maxDim / w, (double)maxDim / h);
        if (scale >= 1.0) return jpegData;

        var scaled = new TransformedBitmap(frame, new System.Windows.Media.ScaleTransform(scale, scale));
        var encoder = new JpegBitmapEncoder { QualityLevel = 70 };
        encoder.Frames.Add(BitmapFrame.Create(scaled));

        using var output = new MemoryStream();
        encoder.Save(output);
        return output.ToArray();
    }
}
